src/data-model-providers/codedriven/CodeDrivenDataModelProvider.cpp

```diff
@@ -0,0 +1,310 @@
+    ReturnErrorOnFailure(epProvider->ServerClusterInterfaces(serverClusterBuilder));
+    auto serverClusters = serverClusterBuilder.TakeBuffer();
+
+    // Ensure capacity before appending.
+    ReturnErrorOnFailure(out.EnsureAppendCapacity(serverClusters.size()));
+
+    for (const auto * serverCluster : serverClusters)
+    {
+        if (!serverCluster->GetPaths().empty())
+        {
+            auto path = serverCluster->GetPaths().front();
+            ReturnErrorOnFailure(
+                out.Append({ path.mClusterId, serverCluster->GetDataVersion(path), serverCluster->GetClusterFlags(path) }));
+        }
+    }
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR CodeDrivenDataModelProvider::GeneratedCommands(const ConcreteClusterPath & path, ReadOnlyBufferBuilder<CommandId> & out)
+{
+    EndpointProviderInterface * epProvider = GetEndpointProvider(path.mEndpointId);
+    if (epProvider == nullptr)
+    {
+        return CHIP_ERROR_NOT_FOUND;
+    }
+    ServerClusterInterface * serverCluster = epProvider->GetServerCluster(path.mClusterId);
+    if (serverCluster == nullptr)
+    {
+        return CHIP_ERROR_NOT_FOUND;
+    }
+    return serverCluster->GeneratedCommands(path, out);
+}
+CHIP_ERROR CodeDrivenDataModelProvider::AcceptedCommands(const ConcreteClusterPath & path,
+                                                         ReadOnlyBufferBuilder<DataModel::AcceptedCommandEntry> & out)
+{
+    EndpointProviderInterface * epProvider = GetEndpointProvider(path.mEndpointId);
+    if (epProvider == nullptr)
+    {
+        return CHIP_ERROR_NOT_FOUND;
+    }
+    ServerClusterInterface * serverCluster = epProvider->GetServerCluster(path.mClusterId);
+    if (serverCluster == nullptr)
+    {
+        return CHIP_ERROR_NOT_FOUND;
+    }
+    return serverCluster->AcceptedCommands(path, out);
+}
+
+CHIP_ERROR CodeDrivenDataModelProvider::Attributes(const ConcreteClusterPath & path,
+                                                   ReadOnlyBufferBuilder<DataModel::AttributeEntry> & out)
+{
+    EndpointProviderInterface * epProvider = GetEndpointProvider(path.mEndpointId);
+    if (epProvider == nullptr)
+    {
+        return CHIP_ERROR_NOT_FOUND;
+    }
+    ServerClusterInterface * serverCluster = epProvider->GetServerCluster(path.mClusterId);
+    if (serverCluster == nullptr)
+    {
+        return CHIP_ERROR_NOT_FOUND;
+    }
+    return serverCluster->Attributes(path, out);
+}
+
+void CodeDrivenDataModelProvider::Temporary_ReportAttributeChanged(const AttributePathParams & path)
+{
+    ChipLogDetail(DataManagement, "CodeDrivenDataModelProvider::Temporary_ReportAttributeChanged");
+}
+
+CHIP_ERROR CodeDrivenDataModelProvider::AddEndpoint(EndpointProviderInterface & endpointProvider)
+{
+    if (GetEndpointProvider(endpointProvider.GetEndpointEntry().id) != nullptr)
+    {
+        return CHIP_ERROR_DUPLICATE_KEY_ID;
+    }
+    // TODO(sergiosoares): Do we need to Startup clusters here?
```

Comment:
I believe yes on both startup and shutdown. However it has to be synchronized with the data model startup:

- if you add an endpoint before the model startup, model startup starts up all those clusters
- if you add an endpoint after the provider startup (e.g. at runtime, a bridge could add a new device endpoint) then startup/shutdown is inline.

Registry does this: https://github.com/project-chip/connectedhomeip/blob/master/src/data-model-providers/codegen/ServerClusterInterfaceRegistry.cpp#L69 and https://github.com/project-chip/connectedhomeip/blob/master/src/data-model-providers/codegen/ServerClusterInterfaceRegistry.cpp#L208